contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -0,0 +1,107 @@
+#include "contrib/hyperscan/matching/input_matchers/source/matcher.h"
+
+#include <numeric>
+
+namespace Envoy {
+namespace Extensions {
+namespace Matching {
+namespace InputMatchers {
+namespace Hyperscan {
+ScratchThreadLocal::ScratchThreadLocal(hs_database_t* database) {
+  hs_error_t err = hs_alloc_scratch(database, &scratch_);
```

Comment:
Is the `hs_alloc_scratch` thread safe? If not you may have a problem here.

I would also add a multithreaded test in config_test.cc. The test will alert you to any thread safety issues during TSAN build. Benchmarks are not being run during CI, and without a multithreaded test you have no idea if the code is correct or not.

Reply:
I am not quite sure the reason. `ScratchThreadLocal` is a thread local object, and each thread will create its own `ScratchThreadLocal`. So I guess there should be no thread-safe concern.

Reply:
`ScratchThreadLocal` constructor is invoked from worker threads, so you may have multiple calls running on multiple threads at the same time. They will share the database pointed by the `database` pointer. If the `hs_alloc_scratch` or some state in the `hs_database_t` object is not thread safe, you may run into problems.

This is why I'm suggesting adding a test with multiple worker threads, so this part of code can be tested. If there are problem that TSAN tests may be able to find it.

Reply:
I catch your point. `hs_alloc_scratch` and `hs_database_t*` are thread safe, but I will add test for `ScratchThreadLocal`.

Reply:
I had missed the fact that there is a shared DB between the TLS instances of ScratchThreadLocal.  Is hs_database_t* actually thread-safe internally? Or are you relying on having it passed in as a const pointer?

That is, if there is a config update and the patterns used for hyperscan change while requests are flowing what's the mechanism used to atomically swap in the new version of the database? This might flow naturally from the TLS infrastructure but I'm not sure. 

https://blog.envoyproxy.io/envoy-threading-model-a8d44b922310 may be useful reading for context.

Reply:
The database is read-only for allocate scratch here.

I believe if there is a config update, the connections in the corresponding filter chain will be drained, and the marcher will be reconstructed and replaced.

Reply:
Refer to https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/listeners/network_filter_chain#filter-chain-only-update, I think there is no race issues.

Reply:
Oh thanks Looong! I had not known about that bit of magic in the infrastructure. Very cool. OK this is fine with me. But I do see in https://intel.github.io/hyperscan/dev-reference/api_files.html#hs__runtime_8h_1ac0ddcd0c2a40f445211e415522fa7078 that hs_alloc_scratch's signature takes a `const hs_database_t*` so why not take one here also?

Reply:
Thanks, I missed it. I will change it to a `const`.

Reply:
Oh, I found that it has become a `const` already. (No idea why GitHub did not mark the file as outdated)